src/game/logic/object/update/physicsupdate.h

```diff
@@ -69,6 +75,21 @@ class PhysicsBehavior : public UpdateModule, public CollideModuleInterface
         return m_velMag;
     }
 
+    float Get_Foward_Speed_2D() const
+    {
+        const Coord3D *dir = Get_Object()->Get_Unit_Dir_Vector2D();
+        float x = m_vel.x * dir->x;
+        float y = m_vel.x * dir->x;
+        float xy = x + y;
+        float len = GameMath::Sqrt(x * x + y * y);
+
+        if (xy < 0.0f) {
+            return -len;
+        } else {
+            return len;
+        }
+    }

```

Comment:
This function or callers of this function could be candidate for this bug:
https://github.com/xezon/GeneralsGamePatch/issues/650